### PR TITLE
Fix marshalling of correlation filter

### PR DESCRIPTION
**Pull request by jhendrixMSFT**
Added custom XML marshaller and unmarshaller for correlation filter rule
with custom properties.

### CODE REVIEW

**Comment by jhendrixMSFT** [comment]
Fixes https://github.com/Azure/azure-service-bus-go/issues/204

File: filter.go [hunk 1]
```diff
@@ -66,3 +76,112 @@ func (cf CorrelationFilter) ToFilterDescription() FilterDescription {
 		CorrelationFilter: cf,
 	}
 }
+
+// MarshalXML implements the xml.Marshaller interface for CorrelationFilterProperties.
+func (c CorrelationFilterProperties) MarshalXML(e *xml.Encoder, start xml.StartElement) error {
```
Comment (jhendrixMSFT) [inline filter.go:81]: Looking closer at the C# impl, I think this needs a bit more work.

**Comment by serbrech** [comment]
Interesting, we used the SQL filter to get around it. that will help.

**Comment by jhendrixMSFT** [comment]
@serbrech I don't suppose you have an easy way to test this?  The C# impl adds various [XML namespace](https://github.com/Azure/azure-sdk-for-net/blob/master/sdk/servicebus/Microsoft.Azure.ServiceBus/src/Filters/XmlObjectConvertor.cs#L99) bits to their serialization which this impl doesn't have at present.

**Comment by serbrech** [comment]
I can write an integration test next week.

File: filter.go (hunk 1 above)
Comment (catalinaperalta) [reply filter.go:81]: What's missing compared to the C# impl?

Comment (jhendrixMSFT) [reply filter.go:81]: There's handling of XML namespace stuffs, see [here](https://github.com/Azure/azure-sdk-for-net/blob/master/sdk/servicebus/Microsoft.Azure.ServiceBus/src/Filters/XmlObjectConvertor.cs#L99).

File: filter.go [hunk 2]
```diff
@@ -66,3 +76,112 @@ func (cf CorrelationFilter) ToFilterDescription() FilterDescription {
...
+		}
+	}
+	return e.EncodeToken(start.End())
+}
+
+// UnmarshalXML implements the xml.Unmarshaller interface for CorrelationFilterProperties.
+func (c *CorrelationFilterProperties) UnmarshalXML(d *xml.Decoder, start xml.StartElement) error {
+	var key string
+	var val interface{}
+	*c = map[string]interface{}{}
```
Comment (catalinaperalta) [inline filter.go:110]: Just double checking, there's no case where c could be nil right?

Comment (jhendrixMSFT) [reply filter.go:110]: Correct.  The unmarshaller will create an instance of that for us.

**Review by catalinaperalta — APPROVED** [review]
LGTM

**Comment by serbrech** [comment]
@jhendrixMSFT wouldn't these be good tests ?
https://github.com/Azure/azure-service-bus-go/blob/9ac8a371a60a98a9bf1304cba94076ce786383fc/subscription_test.go#L325

**Comment by jhendrixMSFT** [comment]
Good catch, I'll take a look.

**Comment by richardpark-msft** [comment]
NOTE: We've moved development of this package to the azure-sdk-for-go repo [link](https://github.com/Azure/azure-sdk-for-go/tree/main/sdk/messaging/azservicebus).

**Pull request closed without merging**